Index: include/clang/Basic/Attr.td
===================================================================
--- include/clang/Basic/Attr.td	(revision 349308)
+++ include/clang/Basic/Attr.td	(working copy)
@@ -1569,6 +1569,12 @@
   let Documentation = [AssumeAlignedDocs];
 }
 
+def EEntry : InheritableAttr {
+  let Spellings = [GCC<"eentry">, Keyword<"eentry">];
+  let Subjects = SubjectList<[Function]>;
+  let Documentation = [Undocumented];
+}
+
 def AllocAlign : InheritableAttr {
   let Spellings = [GCC<"alloc_align">];
   let Subjects = SubjectList<[HasFunctionProto]>;
@@ -1576,6 +1582,7 @@
   let Documentation = [AllocAlignDocs];
 }
 
+
 def NoReturn : InheritableAttr {
   let Spellings = [GCC<"noreturn">, Declspec<"noreturn">];
   // FIXME: Does GCC allow this on the function instead?
Index: lib/CodeGen/CodeGenModule.cpp
===================================================================
--- lib/CodeGen/CodeGenModule.cpp	(revision 349308)
+++ lib/CodeGen/CodeGenModule.cpp	(working copy)
@@ -1569,6 +1569,10 @@
   setLinkageForGV(F, FD);
   setGVProperties(F, FD);
 
+  if (FD->getAttr<EEntryAttr>()) {
+    F->addFnAttr("sllvm-eentry", "true");
+  }
+
   if (const auto *CSA = FD->getAttr<CodeSegAttr>())
     F->setSection(CSA->getName());
   else if (const auto *SA = FD->getAttr<SectionAttr>())
Index: lib/Sema/SemaDeclAttr.cpp
===================================================================
--- lib/Sema/SemaDeclAttr.cpp	(revision 349308)
+++ lib/Sema/SemaDeclAttr.cpp	(working copy)
@@ -6181,6 +6181,9 @@
   case ParsedAttr::AT_IFunc:
     handleIFuncAttr(S, D, AL);
     break;
+  case ParsedAttr::AT_EEntry:
+    handleSimpleAttribute<EEntryAttr>(S, D, AL);
+    break;
   case ParsedAttr::AT_Alias:
     handleAliasAttr(S, D, AL);
     break;
